### ISSUE TITLE
Feature idea: --range

### ISSUE THREAD

**Issue by hoelzro** [beyondgrep/ack3#165]
The new `--range` option would tell ack to only return matches within a range of lines.  For example, if I only wanted to search the first 50 lines of a file, I would do this:

```
$ ack --range 1-50 $search
```

You can specify a range of lines with a dash (as in the example above), or in a comma-separated list (so `--range 1,2,3,4,5` is equivalent to `--range 1-5`).

A sed/perl-like "flipflop" range should also be possible; eg. to search everything but the `DATA` segment of a Perl program:

```
$ ack --perl --range '/./../^__DATA__$/' $search
```

## Additional Thoughts

  * How would `--range` interact with searching multiple files? The flipflop query makes sense, but searching given lines wouldn't always make sense.  On the other hand, I can totally see situations in which it _would_ (eg. searching the first 25 lines of each file for a GPL preamble)
  * `--range` should be easy to use in concert with a command, so you could have a command `perl-sub-ranges` that prints the start and end ranges for each sub in a script, and then you could do `ack --range $(perl-sub-ranges) $query` to only search inside of subs
  * It should be invertible (eg. I should be able to express "search everything _except_ what falls within a POD block)

**Comment by petdance** [comment]
Here's what I've written up so far.  This will go in the manual.

```


    --range-start=PATTERN
    --range-stop=PATTERN
    --range-invert

=head1 RANGE SEARCHING

ack's C<--type> searching lets you specify types of files you want to
search, and you can also specify filenames.  The C<--range-start> and
C<--range-stop> options let you specify ranges of lines to search within
each file.

Say you had the following file:

    # This function calls print on "foo".
    sub foo {
        print 'foo';
    }
    my $print = 1;
    sub bar {
        print 'bar';
    }
    my $task = 'print';

Say you wanted to search for the string "print", but only if it appears in
a subroutine.  Simply calling C<ack print> will give us three matches:

    $ ack print testfile
    # This function calls print on "foo".
        print 'foo';
    my $print = 1;
        print 'bar';
    my $task = 'print';

We can specify ranges of lines that we want ack to search.  The range starts with any line that matches the
pattern C<^sub \w+>, and stops with any line that matches C<^}>.

    $ ack --range-start='^sub \w+' --range-stop='^}' print testfile
        print 'foo';
        print 'bar';

Note that ack searched two ranges of lines.  The listing below shows which
lines were in a range and which were out of the range.

    Out # This function calls print on "foo".
    In  sub foo {
    In      print 'foo';
    In  }
    Out my $print = 1;
    In  sub bar {
    In      print 'bar';
    In  }
    Out my $task = 'print';

You don't have to specify both C<--range-start> and C<--range-stop>.  IF
C<--range-start> is omitted, then the range runs from the first line in the
file unitl the first line that matches C<--range-stop>.  Similarly, if
C<--range-stop> is omitted, the range runs from the first line matching
C<--range-start> to the end of the file.

You can also add the C<--range-invert> option to tell ack to search only
lines that are B<not> in the range.  In our example above, inverting our
range would give us this result.

    $ ack --range-start='^sub \w+' --range-stop='^}' --range-invert print testfile
    # This function calls print on "foo".
    my $print = 1;
    my $task = 'print';

Everything else in ack works the same way.  Using C<-c> option will give a
count of all the matches that appear within those ranges.  The C<-L> will
show only files that do not have matches within the range, and so on.

Using the context options works the same way, and will show context lines
even if those lines fall outside the range.
```

**Comment by petdance** [comment]
I'm starting work on a `range` branch.

**Comment by SparkeyG** [comment]
An additional use case:
`$ ack -i --range-start='insert into' --range-end=')' table_name testfile
my $sql = 'INSERT INTO
table_name VALUES (?,?,?)
`

**Comment by petdance** [comment]
Mind you, the range we're talking about would be a range of lines, not a range of characters within a line.  Does that still fit your thinking?

**Comment by SparkeyG** [comment]
Yes, edited to have multiple lines as I intended.

**Comment by petdance** [comment]
First simple working version of --range-start and --range-end on the range branch.  Lots of special cases have turned up.

```
# -c has to respect ranges.
# -l and -L have to respect ranges.
# -v has to respect ranges.
# Ranges don't affect context.
# What should --passthru do?
# The patterns in --range-start and --range-end have to follow all the flags of the original regex: -Q, -i, etc
# Handle --range-start without --range-end, and vice versa
```

**Comment by n1vux** [comment]
* What should `--passthru` do?

Passthru prints every line irrespective of matches, and that should remain absolute and not be changed by other new features. So should not be affected by Ranges. Ack ranges are not AWK ranges. Our range limits where we match.

We could add a `--passthru-range` or `--passthru=range` that causes our ranges to act as AWK ranges (only print the range) as well as limiting the matching, but that should be separate.

* The patterns in `--range-start` and `--range-end` have to follow all the flags of the original regex: -Q, -i, etc

all or none. I can (and did on slack) argue both sides on this one. 

Since the experienced user can use `(?i)` to make range start and end explicitly insensitive we don't absolutely need flag control on them. (Only our -w isn't a simple Perl RE flag, and for that one should SWYM Say What You Mean on ranges, as in the canonical example.)

The canonical example `--range-start='^sub\s+' --range-end='^}' argues against taking -i from ackrc or -w from commandline.  In theory, -w if applied to the above would have to error, so we need to be able to specify -w for the line pattern and ***not*** for the range patterns.  Since range patterns are highly likely to use BOL/EOL markers, -w is generally contraindicated for them anyway, and user can apply `^` `\b` `$` more intelligently and efficiently than we can.

* Handle --range-start without --range-end, and vice versa

I'm guessing start without end means ignore everything until the first start, and end without start would mean ignore all after.  In both cases, `'__(?:DATA|END)__'` would be a plausible value for all-before or all-after use.

**Comment by petdance** [comment]
In my mind, --range modifies the input file, so I think that `--range-start='^sub {' --range-end='^}' --passthru` would print the contents of every sub in the file.

**Comment by n1vux** [comment]
Our minds differ ... which means ...

- if we only do one way, we need to document the mental-model implemented 
- and we maybe consider supporting both ways

**Comment by petdance** [comment]
> I'm guessing start without end means ignore everything until the first start,

Right, and I've already written that.  I need to put it in docs.

Is there even a case where you would want to use `--range` and `--passthru`?  My standard use of `--passthru` is when tailing a log.

**Comment by n1vux** [comment]
Even with Logs, looking for a "B" but only when between an "A" and a "C" is a plausible thing to do.

**Comment by petdance** [comment]
Done in the just-released 3.1.0.
